### PR TITLE
Update ingestion server removal IP with EC2 approach

### PR DESCRIPTION
**Pull request by stacimc**
## Description

Updates the ingestion server removal IP to reflect changes discussed [here](https://github.com/WordPress/openverse/issues/3925#issuecomment-2207360055) (moving away from the ASG approach and back to managing EC2 instances more directly).

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (`./ov just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`./ov just catalog/generate-docs media-props`
      for the catalog or `./ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by stacimc** [comment]
Adding @sarayourfriend as a reviewer because of infrastructure implications, and since we first discussed this approach! Also pinging @AetherUnbound for the catalog.

**Review by sarayourfriend — APPROVED** [review]
Great! Just two small comments, neither blocking. This is a great simplification and is a good step in the right direction to eventually remove the API from the workers, and just have them immediately start their job when starting. Doing that would make it easy to use spot instances for this work, saving us a huge sum of money on each run.

**Comment by sarayourfriend** [comment]
I've opened a PR with the required infrastructure code change to make this happen: https://github.com/WordPress/openverse-infrastructure/pull/980

File: documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md [hunk 1]
```diff
@@ -426,58 +425,47 @@ refreshes locally and in production).
...
-  [`terminate_instance_in_auto_scaling_group`](https://boto3.amazonaws.com/v1/documentation/api/1.26.86/reference/services/autoscaling/client/terminate_instance_in_auto_scaling_group.html)
-  to terminate the instance. Make sure to set `ShouldDecrementDesiredCapacity`
-  to `True` to ensure that the ASG does not try to replace the instance. This
-  task should use the
+- Finally, use the
+  [`EC2TerminateInstanceOperator`](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/operators/ec2/index.html#airflow.providers.amazon.aws.operators.ec2.EC2TerminateInstanceOperator)
+  to terminate the instance. Make sure to use the
   [`NONE_SKIPPED` TriggerRule](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/dags.html#trigger-rules)
   to ensure that the instances are terminated, even if there are upstream
   failures. (Skips in local env.)
```
[the remark this answers is not in the document]
Comment (stacimc) [reply documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md:463]: It is meant to be indented, good catch!

File: documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md [hunk 2]
```diff
@@ -227,14 +226,11 @@ configuration we can ensure that the latest indexer-worker image is pulled each
 time, which means that any changes to the reindexing code will become live in
 production as soon as the new docker image is published after merging on `main`.
 
-The EC2 approach uses an ASG to achieve a similar result. Because the ASG will
+The EC2 approach uses EC2 operators to achieve a similar result. Because we will
```
Comment (zackkrida) [inline documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md:229]: small question: does "EC2 operators" refer to https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/operators/ec2/index.html?

Comment (stacimc) [reply documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md:229]: It does (there's also one use of an EC2 hook, also from the aws provider). I've linked to each of the specific operators/hooks I'm referring to in the implementation details below! All existing resources.

File: documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md [hunk 3]
```diff
@@ -625,6 +609,13 @@ potential suggestions for that time:
   would involve expanding the indexer workers' API to handle this additional
   task.
 
+### ASG approach
+
+An earlier draft of this implementation plan used an AutoScaling Group for each
+environment. Airflow would use `set_desired_capacity` to set the ASG's capacity,
+and the ASG itself would then spin up (and down) EC2 instances. A disadvantage
+of this approach was that it is impossible to retry a single indexer worker.
+
```
Comment (zackkrida) [inline documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md:618]: Nice, this looks good. I've been meaning to suggest we find a standardized way to denote changes to our IPs and other documents. Perhaps a "changelog" section that links to the PR file diff and summarizes the changes?

What you've done seems sufficient for now, but reminded me to create an issue for this, so thanks :smile: 

https://github.com/WordPress/openverse/issues/4619

**Review by zackkrida — APPROVED** [review]

File: documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md (hunk 2 above)
Comment (zackkrida) [reply documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md:229]: You got to my comment too quickly :laughing: I was just going to delete it after seeing the docs. Thanks!

File: documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md [hunk 4]
```diff
@@ -426,58 +422,47 @@ refreshes locally and in production).
...
-  available in the given environment. Then:
-  - POST to each worker's `reindexing_task` endpoint the `start_index` and
-    `end_index` it should handle
+  available in the given target environment. Then, for each worker:
+  - Use the
+    ['EC2CreateInstanceOperator'](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/operators/ec2/index.html#airflow.providers.amazon.aws.operators.ec2.EC2CreateInstanceOperator)
+    to create a new EC2 instance. The task returns the id of the created
+    instance.
+    - Under the covers this operator is using boto3
+      [`run_instances'](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2/client/run_instances.html).
```
Comment (AetherUnbound) [inline documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md:437]: Nit formatting:
```suggestion
      [`run_instances`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2/client/run_instances.html).
```

File: documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md [hunk 5]
```diff
@@ -426,58 +425,47 @@ refreshes locally and in production).
...
+      We can pass the name or id for the appropriate launch template in the
+      `config` parameter. The launch template names will be hard-coded
+      constants.
+    - Note: it's possible to use this operator to launch all the workers at once
+      by setting the `min_count` and `max_count` parameters. We will instead be
+      using dynamic task mapping to create a separate task for creating each
+      indexer worker individually: this is so that if a single worker fails
+      later in the process, we can retry that worker in isolation.
```
[the remark this answers is not in the document]
Comment (AetherUnbound) [reply documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md:445]: Thanks for making this explicit!

**Review by AetherUnbound — APPROVED** [review]
No objections, only one small formatting suggestion I found while looking at the rendered document. Looks good!

File: documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md (hunk 4 above)
Comment (stacimc) [reply documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md]: Thanks!

**Pull request merged**